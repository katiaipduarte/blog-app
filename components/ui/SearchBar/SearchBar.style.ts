import styled from 'styled-components';

export const SearchBarContainer = styled.form`
    max-width: 32rem;
    padding: 0.5rem 1rem;
    border-radius: 0.4rem;
    border: 1px solid ${({ theme }) => theme.text};
    margin: 1rem 0;

    &:focus-visible,
    &:focus:not(:focus-visible) {
      outline: auto;
      color: ${({ theme }) => theme.secondaryColor};
    }

    input {
      border: none;
      color: ${({ theme }) => theme.text};
      outline: none;
      font-size: 1rem;
      width: calc(100% - 2.5rem);
      margin-left: 0.5rem; 
      background: transparent;
    }

    #search-term-btn,
    #clear-search-term-btn {
      background: none;
      border: none;
      cursor: pointer;
      vertical-align: middle;
      svg {
        color: ${({ theme }) => theme.text};
        height: 1rem;
      }
      &:focus {
        outline: none;
      }
    }
  }

  @media (min-width: 1080px) {
    //   width: 50%;
  }
`;
