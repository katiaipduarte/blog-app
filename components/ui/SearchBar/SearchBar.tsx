import { faSearch, faTimes } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { ChangeEvent, MouseEvent, useState } from 'react';
import { SearchBarContainer } from './SearchBar.style';

type Props = {
  find: (term: string) => void;
};

const SearchBar = (props: Props): JSX.Element => {
  const { find } = props;
  const [searchTerm, setSearchTerm] = useState<string>('');
  const [showClearButton, setShowClearButton] = useState<boolean>(false);

  const handleClearSearch = (): void => {
    setSearchTerm('');
    find('');
    setShowClearButton(false);
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setSearchTerm(event.target.value);
    find(event.target.value);
    setShowClearButton(event.target.value.length !== 0);
  };

  const handleMGBClick = (event: MouseEvent<HTMLButtonElement>): void => {
    event.preventDefault();

    if (searchTerm.length >= 3) {
      onClickSearchOption(searchTerm);
    }
  };

  const onClickSearchOption = (item: string): void => {
    if (item !== '') {
      setSearchTerm(item);
      find(item);
    }
  };

  return (
    <SearchBarContainer
      tabIndex={-1}
      role="searchbox"
      aria-label="Search article by name form"
    >
      <button
        id="search-term-btn"
        aria-label="Click to search"
        onClick={handleMGBClick}
        tabIndex={-1}
      >
        <FontAwesomeIcon icon={faSearch} />
      </button>

      <input
        type="text"
        placeholder="Search article by name"
        value={searchTerm}
        maxLength={50}
        onChange={handleChange}
        role="search"
        aria-label="Search article by name"
      />
      <button
        id="clear-search-term-btn"
        aria-label="Click to clear search"
        onClick={handleClearSearch}
        style={{ visibility: showClearButton ? 'initial' : 'hidden' }}
      >
        <FontAwesomeIcon icon={faTimes} />
      </button>
    </SearchBarContainer>
  );
};

export default SearchBar;
