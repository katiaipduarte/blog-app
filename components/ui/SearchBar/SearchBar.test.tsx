/* eslint-disable @typescript-eslint/ban-types */
import React from 'react';
import SearchBar from './SearchBar';
import 'jest-styled-components';
import { fireEvent, render, screen } from '@testing-library/react';

describe('<SearchBar /> component Unit Test', () => {
  beforeEach(() => {
    render(<SearchBar find={jest.fn().mockName('mockedFunction')} />);
  });

  it('should render Searchbar', () => {
    const { container } = render(
      <SearchBar find={jest.fn().mockName('mockedFunction')} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render 1 <SearchBar /> component', () => {
    expect(screen.queryByRole('searchbox')).toBeInTheDocument();
  });

  it('should not show clear button if input is empty', () => {
    const searchInput = screen.queryByPlaceholderText(
      'Search article by name'
    ) as HTMLInputElement;
    expect(searchInput.value).toBe('');
    expect(
      screen.getByLabelText('Click to clear search').style.visibility
    ).toBe('hidden');
  });

  it('should show clear button if input is not empty', () => {
    const searchInput = screen.queryByPlaceholderText(
      'Search article by name'
    ) as HTMLInputElement;
    fireEvent.change(searchInput, { target: { value: 'test' } });
    expect(searchInput.value).toBe('test');
    expect(
      screen.getByLabelText('Click to clear search').style.visibility
    ).toBe('initial');
  });

  it('should hide the clear button on click', () => {
    const clearBtn = screen.getByLabelText('Click to clear search');
    expect(clearBtn.style.visibility).toBe('hidden');

    const searchInput = screen.queryByPlaceholderText(
      'Search article by name'
    ) as HTMLInputElement;
    fireEvent.change(searchInput, { target: { value: 'test' } });
    expect(clearBtn.style.visibility).toBe('initial');

    fireEvent.click(clearBtn);

    expect(clearBtn.style.visibility).toBe('hidden');
  });

  it('should render an input of type text and allows at max 50 char', () => {
    const searchInput = screen.queryByPlaceholderText(
      'Search article by name'
    ) as HTMLInputElement;
    expect(searchInput.type).toBe('text');
    expect(searchInput.maxLength).toBe(50);

    const validInput = 'ygevhodxbvbaqzpflfkoylzhkbcerwmkavypauvvdwpeunagmq';
    fireEvent.change(searchInput, { target: { value: validInput } });

    expect(validInput.length).toBe(50);
    expect(searchInput.value).toBe(validInput);

    fireEvent.keyDown(searchInput, { key: 'a' });
    expect(searchInput.value).toBe(validInput);
  });
});
