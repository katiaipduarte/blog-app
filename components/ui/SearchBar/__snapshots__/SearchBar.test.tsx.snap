// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> component Unit Test should render Searchbar 1`] = `
.c0 {
  max-width: 32rem;
  padding: 0.5rem 1rem;
  border-radius: 0.4rem;
  border: 1px solid;
  margin: 1rem 0;
}

.c0:focus-visible,
.c0:focus:not(:focus-visible) {
  outline: auto;
}

.c0 input {
  border: none;
  outline: none;
  font-size: 1rem;
  width: calc(100% - 3rem);
  margin-left: 0.5rem;
  background: transparent;
}

.c0 #search-term-btn,
.c0 #clear-search-term-btn {
  background: none;
  border: none;
  cursor: pointer;
  vertical-align: middle;
}

.c0 #search-term-btn svg,
.c0 #clear-search-term-btn svg {
  height: 1rem;
}

.c0 #search-term-btn:focus,
.c0 #clear-search-term-btn:focus {
  outline: none;
}

@media (min-width:1080px) {

}

<form
  aria-label="Search article by name form"
  className="c0"
  role="searchbox"
  tabIndex={-1}
>
  <button
    aria-label="Click to search"
    id="search-term-btn"
    onClick={[Function]}
    tabIndex={-1}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-search fa-w-16 "
      data-icon="search"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
  <input
    aria-label="Search article by name"
    maxLength={50}
    onChange={[Function]}
    placeholder="Search article by name"
    role="search"
    type="text"
    value=""
  />
  <button
    aria-label="Click to clear search"
    id="clear-search-term-btn"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-times fa-w-11 "
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</form>
`;
